src/Algebra/Graph/Algorithm/Bfs.hs

```diff
@@ -0,0 +1,167 @@
+-- |
+-- Breadth-first search (BFS) for directed algebraic graphs and BFS-based algorithms.
+module Algebra.Graph.Algorithm.Bfs where
+
+import           Algebra.Graph
+import qualified Data.Set      as S
+
+-- * BFS core
+
+-- | \( O(?) \).
+-- BFS for algebraic graphs.
+--
+-- Returns list of reachable vertices with corresponding distances
+-- from the starting vertex.
+--
+-- >>> bfs (1 * 2 + 2 * (3 * 4)) 1
+-- [(2,1),(3,2),(4,2)]
+bfs :: Ord a => Graph a -> a -> [(a, Int)]
+bfs graph s = bfsLoop graph s initial S.empty 1
+  where
+    initial = (\x -> (x, 1)) <$> adjacentTo s graph
+
+-- ** Variations
+
+-- | The body of bfs algorithm.
+bfsLoop :: Ord a => Graph a -> a -> [(a, Int)] -> S.Set a -> Int -> [(a, Int)]
+bfsLoop _ _ [] _ _ = []
+bfsLoop graph s (x:xs) visited depth
+  | (S.notMember p visited) && (s /= p) = x : bfsLoop graph s (xs <> conns) (S.insert p visited) (depth + 1)
+  | otherwise = bfsLoop graph s xs visited (depth + 1)
+  where
+    conns = (\v -> (v, depth)) <$> adjacentTo p graph
+    (p, _) = x
+
+-- * Helpers
+
+-- | \( O(s) \).
+-- Find all vertices adjacent to a given one (including the vertex itself).
+--
+-- **NOTE:** may contain duplicates!
+--
+-- >>> adjacentTo 2 (1 * 2 + 2 * 3 * 4)
+-- [2,2,3,4]
+--
+-- Returns empty list if starting point does not belong to the graph:
+--
+-- >>> adjacentTo 5 (1 * 2 + 2 * 3 * 4)
+-- []
+adjacentTo :: Eq a => a -> Graph a -> [a]
+adjacentTo _ Empty = []
+adjacentTo t (Vertex u)
+  | u == t = [t]
+  | otherwise = []
+adjacentTo t (Connect l r)
+  | null leftVertices = adjacentTo t r
+  | otherwise         = leftVertices <> extractVertices r
+  where
+    leftVertices = adjacentTo t l
+adjacentTo t (Overlay l r) = adjacentTo t l <> adjacentTo t r
+
+-- | \( O(s*n) \). Not sure.
+-- Find all reachable vertices a given vertex.
+--
+-- **NOTE:** may contain duplicates and starting point!
+--
+-- >>> reachableFrom 1 (1 * 2 + 2 * 3 + 3 * 4)
+-- [1,2,3,4]
+--
+-- >>> reachableFrom 1 ((1 * 2 + 3 * 4) + (2 * 3))
+-- [1,2,3,3,4]
+reachableFrom :: Eq a => a -> Graph a -> [a]
+reachableFrom t g = t : reachableFromHelper t Empty g
+
+reachableFromHelper :: Eq a => a -> Graph a -> Graph a -> [a]
+reachableFromHelper t Empty (Overlay l r) = left <> right
+  where
+    left = concat $ (\x -> x : reachableFromHelper x Empty r) <$> reachableFromHelper t r l
```

Comment:
Here you are traversing entire right graph for **every** reachable vertex from the left graph. Consider a graph like this:
<img width="1117" alt="Screenshot 2020-06-11 at 08 40 54" src="https://user-images.githubusercontent.com/686582/84349107-48667980-abbf-11ea-939a-31f22b3c78ee.png">

If you compute `reachableFrom 1` for this graph, you will find 3 reachable vertices (without starting one) in the first biclique, then you will 3 more reachable vertices in the second one **for every one of the 3 vertices before**, so this will yield 3 + 9 vertices total. Then you will find 3 more reachable vertices, but you will find them for every of the 9 occurrences in the previous biclique.

You can see that the complexity here easily becomes exponential. And you can test it like this (for a graph with 12 bicliques connected to each other):

```haskell
>>> let g n = overlays [ biclique [3*k+1..3*k+3] [3*k+4..3*k+6] | k <- [0..n-1]]
>>> :set +s
>>> size (g 12)
72
(0.00 secs, 453,600 bytes)
>>> length (reachableFrom 1 (g 12))
797161
(2.35 secs, 2,624,686,776 bytes)
```

Here are possible ways to improve:
1. Remove duplicates found on the left before recursing on the right (same for right to left). Note that removing duplicates is not linear in general (it can be quadratic for `Eq` or O(n log n) for `Ord`). However, for `Int`s this should make the time complexity of `reachableFrom` close to O(s*n).
2. Generalize the function to compute reachable vertices from a given list of vertices. You are doing this anyway already in `Overlay`, but you will change implementation for other constructors. This step should ensure that you are not traversing right/left graph multiple times (well, an easy solution is two traverse left/right at most twice). This should bring the time complexity down to something like O(s + n*k) where k is the number of vertices you start from.
3. Instead of computing reachable from left graph before looking at the right graph and then looking at the left graph again... we should ideally extract something from both graphs that we can "merge" without looking at any part of the graph twice. If you only consider a single set of reachable vertices, then it is not enough. But if you know sets of reachable vertices for all vertices in both graphs (e.g. using `Map v (Set v)`) then you could easily merge those in linear time. Of course now we will be solving a more general problem: **for every vertex** find all vertices reachable from it. But instead of O(s + n^2) it should become something like O(s) or O(s + n log n).